### PR TITLE
Explosive balancing and fixing.

### PR DESCRIPTION
**Pull request by Zukulini**
#### Summary

SUMMARY: Content "Fixes and rebalances to the available craftable explosives."

#### Purpose of change

To make the available explosive recipes more realistic and intuitive.

#### Describe the solution

I noticed that in the game there was a number of size category for explosives depending on the internal size of their containers:
Grenade - 250 ml
Pipe Bomb - 445 ml (though the container itself replaces the shrapnel introduced in other recipes, which makes it roughly equivalent to a grenade with an internal size of 500 ml)
Small Frag Device - 3750 ml
Demolition Charge - 3750 ml (it does not introduce any metal that might produce shrapnel)
Barrel Bomb - 100L

These categories are paired with adjectives which correlate to the explosives they utilize:
Primitive - Utilizes only volatile explosives.
Improvised - Utilizes mostly stable explosives, and a minimal ammount of volatile explosives to initiate the detonation (trending towards a ratio of 1 volatile to 39 stable).
Bootleg - Utilizes military explosives (the only explosive in this requirement is Composition B).

Curiously, all items in the same size category had the exact same explosive power, all grenades had a power of 240, and so did the Small Improvised Frag Device and Small Bootleg Frag Device have equivalent powers, even though Concentration B is much more powerful as per its volume than any other Volatile or Stable explosives.

The ammount of explosive powders used in each recipe was out of proportion to the size of their container, some leaning towards having too much (the Grenades) and some leaning towards having too little (the Small Frag Devices, the Demolition Charges and the Barrel Bombs). To fix this, I measured how much volume a portion the **least dense per explosive power** (the ones that took up the most volume to produce an explosion of power 10) options of Volatile (ammonium nitrate pellets), Stable (HMTD) and Military (Composition B) would take. Then I remade the recipes for each explosive so that it's possible to fill the container to the brim, but impossible to overfill it with explosive powders. 

I concluded that Primitive and Improvised explosives were too redundant, so I merged them both into Homemade explosives, which had their values fixed as explained in the previous paragraph. I made two recipes for some Homemade explosives, that allow them to be made with either Volatile and Stable explosives or only Volatile explosives. After having some discussions in the development discord server, I realized that it would only be realistic to allow the smallest explosives to utilize only volatile explosives, because an explosion like that requires a hard case and because having large concentrations of volatile explosives in one container alone would mean an extremely high chance of having random friction or static ignite the entire system; and thus I restricted that option to only grenades and pipe bombs.

I also made a different set of explosives with "Military Explosive" to replace the ones with the "Bootleg" adjective. I made these with the same principles as Homemade explosives, so that you can completly fill the containers with Composition B. Since Composition B is more powerful per volume, this means that Military Explosives are more powerful than their Homemade counterparts.

Lastly, I made 2 other size categories of explosives, 500 ml grenades (the original 250 ml grenades are renamed as "small grenades", as they can be made with small tin cans, and "grenades" are made with medium tin cans) and 50L barrel bombs, which use Steel Kegs. Both of these categories have Homemade and Military Explosive variants.

Also a couple of small, related QoL additions: I fixed the density of match head powder, which used to be as dense as water when it should be around twice as heavy, thusly fixed its number on the volatile explosive requirement list to reflect its new weight, and also introduced a new requirement list with 250 ml containers that can be used in various grenades.    

#### Describe alternatives you've considered

I considered making a separate category for explosives which utilized only volatile powders, which would have overall a higher power per volume. However I thought that it would be realistic to have volatile only explosives not be tightly packed, and didn't want to clog up the list with more redundant items.

#### Testing

I tried my explosives in a debug mode world, and they seem to be working as intended. 

#### Additional context

Currently, loading a previous version save with these changes will provoke a set of skippable errors like these on the final loading text:

 FUNCTION : const T& string_id<T>::obj() const [with T = recipe]
 FILE     : src/recipe_dictionary.cpp
 LINE     : 65
 VERSION  : bea0650
DEBUG    : invalid recipe id "(obsoleted recipe)"

I would appreciate any help in getting rid of these error messages.

### CODE REVIEW

File: data/json/recipes/weapon/explosive.json [hunk 1]
```diff
@@ -333,6 +333,7 @@
   },
   {
     "result": "small_homemade_grenade",
+    "id_suffix": "volatile"
```
Comment (Wishbringer) [inline data/json/recipes/weapon/explosive.json:336]:
```suggestion
    "id_suffix": "volatile",
```

File: data/json/recipes/weapon/explosive.json [hunk 2]
```diff
@@ -372,6 +373,7 @@
   },
   {
     "result": "homemade_grenade",
+    "id_suffix": "volatile"
```
Comment (Wishbringer) [inline data/json/recipes/weapon/explosive.json]:
```suggestion
    "id_suffix": "volatile",
```

File: data/json/recipes/weapon/explosive.json [hunk 3]
```diff
@@ -425,6 +427,7 @@
   },
   {
     "result": "pipebomb",
+    "id_suffix": "volatile"
```
Comment (Wishbringer) [inline data/json/recipes/weapon/explosive.json]:
```suggestion
    "id_suffix": "volatile",
```

**Comment by RenechCDDA** [comment]
> Currently, loading a previous version save with these changes will provoke a set of skippable errors like these on the final loading text:
> 
> FUNCTION : const T& string_id::obj() const [with T = recipe]
> FILE : src/recipe_dictionary.cpp
> LINE : 65
> VERSION : [bea0650](https://github.com/CleverRaven/Cataclysm-DDA/commit/bea06502906d98d729d84876e82406c67897f348)
> DEBUG : invalid recipe id "(obsoleted recipe)"
> 
> I would appreciate any help in getting rid of these error messages.

Went through to check them all, they're all pretty much the same thing.

Obsoleted recipes don't go into `data/json/obsolete.json`. They go into `data/json/recipes/recipe_obsolete.json`.

That applies to the following recipes:

improvised_demolition_charge
bootleg_grenade 
primitive_demolition_charge 
tool_bootleg_barrel_bomb 
tool_small_bootleg_fragmentation_device
primitive_grenade
improvised_pipebomb
primitive_pipebomb
tool_improvised_barrel_bomb
improvised_grenade
can_bomb
tool_small_improvised_fragmentation_device

Basically just migrate all of your entries in `data/json/obsolete.json ` to `data/json/recipes/recipe_obsolete.json` using the format you see in `recipe_obsolete.json`.

The only one you need to watch out for is bootleg_pipebomb_stable, as that has a suffix (applied outside of the `id` field)

**Comment by RenechCDDA** [comment]
@Zukulini I don't have the access required to make the changes on this PR directly, but I put a PR in for your PR's branch. So if  you just want to easily resolve those errors with two clicks you can. 

You can also independently resolve them as outlined above, if you wish. I do not care about "credit" so feel free to copy without attribution.

https://github.com/Zukulini/Cataclysm-DDA/pull/1

**Comment by Zukulini** [comment]
@RenechCDDA thanks a lot, I just merged your PR!

**Comment by cainiaowu** [comment]
duplicate entry of pipebomb in https://github.com/CleverRaven/Cataclysm-DDA/blob/515f804b4aadbaa70211507f948e23094a9ecc8f/data/json/obsolete.json#L878
